annotation/tests/test_annotations.py

```diff
@@ -482,3 +593,665 @@ def test_construct_annotated_doc_db_error(
+    annotation_doc_for_save.pages = None
+    annotation_doc_for_save.validated = None
+    annotation_doc_for_save.failed_validation_pages = None
+
+    check_null_fields(annotation_doc_for_save)
+    assert annotation_doc_for_save.pages == []
+    assert annotation_doc_for_save.validated == set()
+    assert annotation_doc_for_save.failed_validation_pages == set()
+
+
+@pytest.mark.parametrize(("stop_revision"), ("not found", None))
+def test_accumulate_pages_info_stop_rev_none(
+    stop_revision: str, annotated_doc: AnnotatedDoc
+):
+    valid, fail, annot, not_proc, categ, required_rev = accumulate_pages_info(
+        task_pages=[1, 2, 3],
+        revisions=[annotated_doc],
+        stop_revision=stop_revision,
+        specific_pages=False,
+        with_page_hash=False,
+        unique_status=False,
+    )
+    assert valid == {1}
+    assert fail == set()
+    assert annot == set()
+    assert not_proc == {2, 3}
+    assert categ == annotated_doc.categories
+    assert required_rev is None
+
+
+@pytest.mark.parametrize(
+    ("stop_revision"), (LATEST, "20fe52cce6a632c6eb09fdc5b3e1594f926eea69")
+)
+def test_accumulate_pages_info_stop_rev_not_none(
+    stop_revision: str, annotated_doc: AnnotatedDoc
+):
+    valid, fail, annot, not_proc, categ, required_rev = accumulate_pages_info(
+        task_pages=[1, 2, 3],
+        revisions=[annotated_doc],
+        stop_revision=stop_revision,
+        specific_pages=False,
+        with_page_hash=False,
+        unique_status=False,
+    )
+    assert valid == {1}
+    assert fail == set()
+    assert annot == set()
+    assert not_proc == {2, 3}
+    assert categ == annotated_doc.categories
+    assert required_rev == annotated_doc
+
+
+def test_accumulate_pages_info_specific_pages_hash(
+    annotated_doc: AnnotatedDoc,
+):
+    # annotated_doc.pages["1"] = "sha1"
+    annotated_doc.pages["2"] = "sha2"
+    valid, fail, annot, not_proc, categ, required_rev = accumulate_pages_info(
+        task_pages=[1, 2, 3],
+        revisions=[annotated_doc],
+        stop_revision=None,
+        specific_pages={2},
+        with_page_hash=True,
+        unique_status=False,
+    )
+    assert valid == set()
+    assert fail == set()
+    assert annot == {"2": "sha2"}
+    assert not_proc == {1, 2, 3}
+    assert categ == annotated_doc.categories
+    assert required_rev is None
+
+
+def test_accumulate_pages_info_unique_status(annotated_doc: AnnotatedDoc):
+    annotated_doc.pages["2"] = "sha2"
+    annotated_doc.validated.append(2)
+
+    valid, fail, annot, not_proc, categ, required_rev = accumulate_pages_info(
+        task_pages=[1, 2, 3],
+        revisions=[annotated_doc],
+        stop_revision=None,
+        specific_pages=None,
+        with_page_hash=False,
+        unique_status=True,
+    )
+    assert valid == {1, 2}
+    assert fail == set()
+    assert annot == set()
+    assert not_proc == {3}
+    assert categ == annotated_doc.categories
+    assert required_rev is None
```

Comment:
Since the last three functions are similar, you might want to consider extracting the common part into a separate fixture or writing a single parameterized test.

Reply:
I decided on keeping them as separate tests, because I would need to add logic to tests if I used parametrized test.